types/styled-components/index.d.ts

```diff
@@ -48,82 +69,183 @@ export interface Styles {
+    // NOTE: TS <3.2 will refuse to infer the generic and this component becomes impossible to use in JSX
+    // just the presence of the overload is enough to break JSX
+    //
+    // TODO (TypeScript 3.2): actually makes the 'as' prop polymorphic
+    // (
+    //     props: StyledComponentProps<C, T, O, A> & { as?: never }
+    //   ): React.ReactElement<StyledComponentProps<C, T, O, A>>
+    // <AsC extends keyof JSX.IntrinsicElements | React.ComponentType<any> = C>(
+    //   props: StyledComponentPropsWithAs<AsC, T, O, A>
+    // ): React.ReactElement<StyledComponentPropsWithAs<AsC, T, O, A>>
+
+    // TODO (TypeScript 3.2): delete this overload
+    (
+        props: StyledComponentProps<C, T, O, A> & {
+            /**
+             * Typing Note: prefer using .withComponent for now as it is actually type-safe.
+             *
+             * String types need to be cast to themselves to become literal types (as={'a' as 'a'}).
+             */
+            as?: keyof JSX.IntrinsicElements | React.ComponentType<any>
+        }
+    ): React.ReactElement<StyledComponentProps<C, T, O, A>>;
+    withComponent<WithC extends AnyStyledComponent>(
+        component: WithC
+    ): StyledComponent<
+        StyledComponentInnerComponent<WithC>,
         T,
-        JSX.IntrinsicElements[K] & O
+        O & StyledComponentInnerOtherProps<WithC>,
+        A | StyledComponentInnerAttrs<WithC>
     >;
-    withComponent<U = {}>(
-        element: React.ComponentType<U>,
-    ): StyledComponentClass<U, T, U & O>;
+    withComponent<
+        WithC extends keyof JSX.IntrinsicElements | React.ComponentType<any>
+    >(
+        component: WithC
+    ): StyledComponent<WithC, T, O, A>;
 }
 
-export interface ThemedStyledFunction<P, T, O = P> {
+export interface ThemedStyledFunction<
+    C extends keyof JSX.IntrinsicElements | React.ComponentType<any>,
+    T extends object,
+    O extends object = {},
+    A extends keyof any = never
+> {
     (
         strings: TemplateStringsArray,
-        ...interpolations: Array<Interpolation<ThemedStyledProps<P, T>>>
-    ): StyledComponentClass<P, T, O>;
-    <U>(
+        ...interpolations: Array<
+            Interpolation<
+                ThemedStyledProps<StyledComponentPropsWithRef<C> & O, T>
+            >
+        >
+    ): StyledComponent<C, T, O, A>;
+    <U extends object>(
         strings: TemplateStringsArray,
-        ...interpolations: Array<Interpolation<ThemedStyledProps<P & U, T>>>
-    ): StyledComponentClass<P & U, T, O & U>;
-    attrs<U, A extends Partial<P & U> & { [others: string]: any; } = {}>(
-        attrs: Attrs<P & U, A, T>,
-    ): ThemedStyledFunction<DiffBetween<A, P & U>, T, DiffBetween<A, O & U>>;
+        ...interpolations: Array<
+            Interpolation<
+                ThemedStyledProps<StyledComponentPropsWithRef<C> & O & U, T>
+            >
+        >
+    ): StyledComponent<C, T, O & U, A>;
```

Comment:
One potential breaking change I noticed in my own code is that `U`'s inference seems to be more eager.

If you have two function interpolations, and only the second one annotates its props, `U` is still inferred as `{}` instead of as the annotated type of the second interpolation. The first interpolation is the one that needs its props.